### PR TITLE
[WIP] Migrate EC2 / TGW Route Table to AWS SDK v2

### PR DESCRIPTION
**Pull request by sc250024**
## Description

Fixes #770.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [X] Update the docs.
- [X] Run the relevant tests successfully, including pre-commit checks.
- [X] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.
- [x] Attention Grunts - if this PR adds support for a new resource, ensure the `nuke_sandbox` and `nuke_phxdevops` jobs in `.circleci/config.yml` have been updated with appropriate exclusions (either directly in the job or via the `.circleci/nuke_config.yml` file) to prevent nuking IAM roles, groups, resources, etc that are important for the test accounts.

## Release Notes (draft)

Updated `ec2` and `transit-gateway-route-table` to AWS SDK v2

### CODE REVIEW

File: aws/resources/ec2.go [hunk 1]
```diff
@@ -140,722 +130,34 @@ func (ei *EC2Instances) nukeAll(instanceIds []*string) error {
...
 	if err != nil {
 		logging.Debugf("[Failed] %s", err)
 		return errors.WithStackTrace(err)
 	}
 
-	logging.Debugf("[OK] %d instance(s) terminated in %s", len(instanceIds), ei.Region)
-	return nil
-}
-
-func GetEc2ServiceClient(region string) ec2iface.EC2API {
```
Comment (sc250024) [inline aws/resources/ec2.go:173]: To be honest, I was not sure why there is everything below this line when there is already a separate `ec2_vpc.go` file?

File: aws/resources/ec2.go [hunk 2]
```diff
@@ -140,722 +130,34 @@ func (ei *EC2Instances) nukeAll(instanceIds []*string) error {
...
-
-	return nil
-}
-
-func (v Vpc) nukeEgressOnlyGateways(spinner *pterm.SpinnerPrinter) error {
-	allEgressGateways := []*string{}
-	msg := "Finding Egress Only Internet Gateways to Nuke"
-	spinner.UpdateText(msg)
-	logging.Debug(msg)
-	err := v.svc.DescribeEgressOnlyInternetGatewaysPages(
```
Comment (sc250024) [inline aws/resources/ec2.go:498]: I was not sure how to implement the `paginator` for this if we don't have access to `ei.Client`

File: aws/resources/ec2.go [hunk 3]
```diff
@@ -140,722 +130,34 @@ func (ei *EC2Instances) nukeAll(instanceIds []*string) error {
...
+	_, err = ei.Client.TerminateInstances(ei.Context, params)
 	if err != nil {
 		logging.Debugf("[Failed] %s", err)
 		return errors.WithStackTrace(err)
 	}
 
-	err = ei.Client.WaitUntilInstanceTerminatedWithContext(ei.Context, &ec2.DescribeInstancesInput{
-		Filters: []*ec2.Filter{
+	waiter := ec2.NewInstanceTerminatedWaiter(ei.Client)
+	err = waiter.Wait(ei.Context, &ec2.DescribeInstancesInput{
```
Comment (sc250024) [inline aws/resources/ec2.go:143]: The tests fail at this line, and I'm not sure what it's complaining about. Ideas?

<details>
<summary>Details</summary>

```
--- FAIL: TestEc2InstancesWithEIP_NukeAll (0.00s)
panic: cannot create context from nil parent [recovered]
	panic: cannot create context from nil parent

goroutine 125 [running]:
testing.tRunner.func1.2({0x10375f7c0, 0x103ba6620})
	/opt/homebrew/Cellar/go/1.23.3/libexec/src/testing/testing.go:1632 +0x1bc
testing.tRunner.func1()
	/opt/homebrew/Cellar/go/1.23.3/libexec/src/testing/testing.go:1635 +0x334
panic({0x10375f7c0?, 0x103ba6620?})
	/opt/homebrew/Cellar/go/1.23.3/libexec/src/runtime/panic.go:785 +0x124
context.WithDeadlineCause({0x0?, 0x0?}, {0x1041e9e20?, 0xd18c2e2800?, 0x1041e9e20?}, {0x0?, 0x0?})
	/opt/homebrew/Cellar/go/1.23.3/libexec/src/context/context.go:620 +0x368
context.WithDeadline(...)
	/opt/homebrew/Cellar/go/1.23.3/libexec/src/context/context.go:612
context.WithTimeout({0x0, 0x0}, 0xd18c2e2800)
	/opt/homebrew/Cellar/go/1.23.3/libexec/src/context/context.go:690 +0x50
github.com/aws/aws-sdk-go-v2/service/ec2.(*InstanceTerminatedWaiter).WaitForOutput(0x14000870420, {0x0, 0x0}, 0x14000828280, 0xd18c2e2800, {0x0, 0x0, 0xb0?})
	/my/path/to/go/pkg/mod/github.com/aws/aws-sdk-go-v2/service/ec2@v1.193.0/api_op_DescribeInstances.go:1428 +0x17c
github.com/aws/aws-sdk-go-v2/service/ec2.(*InstanceTerminatedWaiter).Wait(...)
	/my/path/to/go/pkg/mod/github.com/aws/aws-sdk-go-v2/service/ec2@v1.193.0/api_op_DescribeInstances.go:1402
github.com/gruntwork-io/cloud-nuke/aws/resources.(*EC2Instances).nukeAll(0x14000a43dd8, {0x14000a43e98, 0x1, 0x102ea78e0?})
	/my/path/to/cloud-nuke/aws/resources/ec2.go:143 +0x454
github.com/gruntwork-io/cloud-nuke/aws/resources.TestEc2InstancesWithEIP_NukeAll(0x14000449040)
	/my/path/to/cloud-nuke/aws/resources/ec2_test.go:177 +0x178
testing.tRunner(0x14000449040, 0x103ba4398)
	/opt/homebrew/Cellar/go/1.23.3/libexec/src/testing/testing.go:1690 +0xe4
created by testing.(*T).Run in goroutine 1
	/opt/homebrew/Cellar/go/1.23.3/libexec/src/testing/testing.go:1743 +0x314
```

</details>

Comment (wakeful) [reply aws/resources/ec2.go:143]: Did you try to provide `context` when creating the `mock Client` in your test? e.q.

```go
ei := EC2Instances{
		BaseAwsResource: BaseAwsResource{
			Context: context.Background(),
		},
		Client: mockedEC2Instances{
		...
```

File: aws/resources/ec2.go (hunk 1 above)
Comment (wakeful) [reply aws/resources/ec2.go:173]: I actually think this is some old or dead code,
That function is called from `GetDefaultSecurityGroups` and `NewVpcPerRegion`, but I am unable to locate calls to them. This likely needs to be cleaned up or removed, or the signatures should be fixed to allow the code to build.

File: aws/resources/ec2.go (hunk 2 above)
Comment (wakeful) [reply aws/resources/ec2.go:498]: by looking at the code you should use the client under `svc` on the `VPC type` e.q.

```go
	v.svc.ClientAPI_Call
```

but first you also need to update the Client on the **VPC** type with **new sdkV2** check line `180`

File: aws/resources/ec2.go (hunk 1 above)
Comment (sc250024) [reply aws/resources/ec2.go:173]: @wakeful Yes, that's what I figured. In Goland I see something like the following for many of those functions:

Screenshot 2024-11-28 at 06 25 25

These functions are greyed out:

- NewVpcPerRegion
- GetDefaultVpcs
- NukeVpcs
- GetDefaultSecurityGroups
- NukeDefaultSecurityGroupRules

Then all of the `nuke*` functions are called by the main `(v Vpc) nuke` function in that file. These `nuke*` functions also exist in other places. For example:

- `nukeInternetGateways` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L325
- `nukeEgressOnlyGateways` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L357
- `nukeEndpoints` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L393
- `nukeNetworkInterfaces` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L433
- `nukeSubnets` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L474
- `nukeRouteTables` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L539
- `nukeNacls` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L600
- `nukeSecurityGroups` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L656
- `dissociateDhcpOptions` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L738
- `nukeVpc` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L774

That's why I am wondering what's the overall strategy of having this logic in multiple places?

Comment (wakeful) [reply aws/resources/ec2.go:173]: If the linter finds a `func` that is not being called, it is marked as greyed out/unused. I suspect this is leftover code from a previous refactor. IMHO, you can safely remove it if it is not being called.

Comment (james03160927) [reply aws/resources/ec2.go:173]: Thanks for refactoring the code. Can you have a separate PR to address the refactoring code and make this just focus on migrating to sdk-v2 so we can easily tell the issue if something comes up in the future? 

Just CC me for review. I can quickly go over the change. Runnign the unit tests should be able to verify if those are safe to get rid of.

File: aws/resources/ec2.go (hunk 3 above)
Comment (sc250024) [reply aws/resources/ec2.go:143]: Yes, this worked! Thank you @wakeful

File: aws/resources/ec2.go (hunk 1 above)
Comment (sc250024) [reply aws/resources/ec2.go:173]: Hey @james03160927. I've spun off the following two PRs:

- https://github.com/gruntwork-io/cloud-nuke/pull/801
- https://github.com/gruntwork-io/cloud-nuke/pull/803

The https://github.com/gruntwork-io/cloud-nuke/pull/801 PR cherry-picks the commits from the PR that related to the TGW route tables. The https://github.com/gruntwork-io/cloud-nuke/pull/803 PR removes all of the unneeded EC2 code.

I will rebase this PR once those two are merged.

File: aws/resources/ec2_types.go [hunk 4]
```diff
@@ -3,26 +3,56 @@ package resources
...
-	"github.com/aws/aws-sdk-go/service/ec2"
-	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+
 	"github.com/gruntwork-io/cloud-nuke/config"
 	"github.com/gruntwork-io/go-commons/errors"
 )
 
+type EC2InstancesAPI interface {
```
[the remark this answers is not in the document]
Comment (sc250024) [reply aws/resources/ec2_types.go:13]: Is there another interface I should be using instead?

**Comment by james03160927** [comment]
Will close this PR as it has been resolved in other PRs.

**Comment by sc250024** [comment]
@james03160927 I think this needs to remain open. There's still the `ec2.go` file to migrate. Or are you doing that somewhere else?

**Comment by james03160927** [comment]
Ah okay. Sorry about that. Feel free to reopen it and rebase it to the master branch!

**Comment by wakeful** [comment]
hey @sc250024 

Will you have some time over the next few days to recreate your changes and push them for review, or should I pick up the last part of the migration?

**Pull request closed without merging**